sklearn/compose/_column_transformer.py

```diff
@@ -759,3 +761,54 @@ def is_neg(x): return isinstance(x, numbers.Integral) and x < 0
     elif _determine_key_type(key) == 'int':
         return np.any(np.asarray(key) < 0)
     return False
+
+
+class _ColumnSelector:
+    """Callable to select columns. Returned by `make_column_selector`"""
+
+    def __init__(self, pattern, dtype_include, dtype_exclude):
+        self.pattern = pattern
+        self.dtype_include = dtype_include
+        self.dtype_exclude = dtype_exclude
+
+    def __call__(self, df):
+        if not hasattr(df, 'iloc'):
+            raise ValueError("make_column_selector can only be applied to "
+                             "pandas dataframes")
+        if self.dtype_include is not None or self.dtype_exclude is not None:
+            cols = (df.iloc[:1].select_dtypes(include=self.dtype_include,
+                                              exclude=self.dtype_exclude)
+                    .columns)
+        else:
+            cols = df.columns
+        if self.pattern is not None:
+            return cols[cols.str.contains(self.pattern)]
+        return cols
+
+
+def make_column_selector(pattern=None, dtype_include=None, dtype_exclude=None):
+    """Create a callable to select columns to be used with
+    :func:`make_column_transformer` or :class:`ColumnTransformer`.
+
+    :func:`make_column_selector` can select columsn based on a columns datatype
+    or a regex pattern.
+
+    Parameters
+    ----------
+    pattern : str, default=None
+        Regular expression used to select columns. If None, column selection
```

Comment:
Not clear that any column containing this pattern, rather than entirely matching, will be returned

Reply:
Not clear that columns will only be returned if they match *all* criteria